src/Microsoft.AspNet.Mvc.Core/Internal/FilterCache.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNet.Mvc.Abstractions;
+using Microsoft.AspNet.Mvc.Filters;
+using Microsoft.AspNet.Mvc.Infrastructure;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class FilterCache
+    {
+        private readonly IFilterMetadata[] EmptyFilterArray = new IFilterMetadata[0];
+
+        private readonly IActionDescriptorCollectionProvider _collectionProvider;
+        private readonly IFilterProvider[] _filterProviders;
+
+        private InnerCache _currentCache;
+
+        public FilterCache(
+            IActionDescriptorCollectionProvider collectionProvider,
+            IEnumerable<IFilterProvider> filterProviders)
+        {
+            _collectionProvider = collectionProvider;
+            _filterProviders = filterProviders.OrderBy(item => item.Order).ToArray();
+        }
+
+        private InnerCache CurrentCache
+        {
+            get
+            {
+                var current = Volatile.Read(ref _currentCache);
```

Comment:
Yeah, I can switch this if it's going to be easier to grok. I read this http://joeduffyblog.com/2010/12/04/sayonara-volatile/ not too long ago, which lead me to prefer this style. I can see it's causing you pain though.